src/asset-swapper/utils/market_operation_utils/path_optimizer.ts

```diff
@@ -35,65 +35,276 @@ function nativeOrderToNormalizedAmounts(
     return { input, output };
 }
 
-function calculateOutputFee(
-    side: MarketOperation,
-    sampleOrNativeOrder: DexSample | NativeOrderWithFillableAmounts,
-    outputAmountPerEth: BigNumber,
-    inputAmountPerEth: BigNumber,
-    fees: FeeSchedule,
-): BigNumber {
-    if (isDexSample(sampleOrNativeOrder)) {
-        const { input, output, source, fillData } = sampleOrNativeOrder;
-        const fee = fees[source]?.(fillData).fee || ZERO_AMOUNT;
-        const outputFee = ethToOutputAmount({
-            input,
-            output,
-            inputAmountPerEth,
-            outputAmountPerEth,
-            ethAmount: fee,
-        });
-        return outputFee;
-    } else {
-        const { input, output } = nativeOrderToNormalizedAmounts(side, sampleOrNativeOrder);
-        const fee = fees[ERC20BridgeSource.Native]?.(sampleOrNativeOrder).fee || ZERO_AMOUNT;
-        const outputFee = ethToOutputAmount({
-            input,
-            output,
-            inputAmountPerEth,
-            outputAmountPerEth,
-            ethAmount: fee,
+export interface PathOptimizerContext {
+    side: MarketOperation;
+    chainId: ChainId;
+    feeSchedule: FeeSchedule;
+    neonRouterNumSamples: number;
+    fillAdjustor: FillAdjustor;
+    pathPenaltyOpts: PathPenaltyOpts;
+    inputAmount: BigNumber;
+}
+
+export class PathOptimizer {
+    private side: MarketOperation;
+    private chainId: ChainId;
+    private feeSchedule: FeeSchedule;
+    private neonRouterNumSamples: number;
+    private fillAdjustor: FillAdjustor;
+    private pathPenaltyOpts: PathPenaltyOpts;
+    private inputAmount: BigNumber;
+
+    constructor(context: PathOptimizerContext) {
+        this.side = context.side;
+        this.chainId = context.chainId;
+        this.feeSchedule = context.feeSchedule;
+        this.neonRouterNumSamples = context.neonRouterNumSamples;
+        this.fillAdjustor = context.fillAdjustor;
+        this.pathPenaltyOpts = context.pathPenaltyOpts;
+        this.inputAmount = context.inputAmount;
+    }
```

Comment:
For the future: wondering if we can construct a `PathOptimizer` singleton and have some of these fields get passed in at `findOptimalPathFromSamples`. For example, some things can probably be constructed just once:
- `chainId`
- `feeSchedule`

A couple of these items, I'm not sure if they are dynamic (i.e. `fillAdjustor`, `pathPenaltyOpts`)

Reply:
Yeah, I think many fields are pretty much constants (some are like options that are never overridden -- leftover cleanup work from asset-swapper migration and I think many can be removed and replaced with a constant/singleton). 

I am also starting to think some grouped data object (e.g. quote/request context) can be useful.